### DISCUSSION TITLE
Computing a weighted rolling average

### DISCUSSION THREAD

**Discussion by corakingdon** [pydata/xarray#6761] [Q&A] [answered]
I am trying to compute a weighted rolling average where each data point has a different weight. (My example is that I have monthly data, and I want each rolling window average to use the number of days in each month as a weight). The following code works, but I wanted to check if there is a better way to do it that doesn't rely on manual iteration of the DataArrayRolling objects.

```python
import xarray as xr
import numpy as np

window_size = 3

# Create one year of monthly data (twelve data points)
da = xr.DataArray(np.arange(0.,12.), coords={"time":np.arange("2022-01", "2023-01", dtype='datetime64[M]')})
month_weights = da.time.dt.daysinmonth # The numbers of days in each month are the desired weights

# Compute a weighted rolling average with manual iteration
da_rolling = da.rolling(time=window_size)
month_weights_rolling = month_weights.rolling(time=window_size)

result = da.copy()
for (da_label, da_window), (_, weights_window),  in zip(da_rolling, month_weights_rolling):
    result.loc[{"time" : da_label}] = da_window.weighted(weights_window.fillna(0)).mean("time")
```
This is the only way I figured out how to do a rolling average that uses weights unique to each data point (rather than weights that remain the same for each window, which is possible using `construct`).

I thought something like the following might be available:
```python
result = da.weighted(month_weights).rolling(time = window_size).mean("time")
```
but DataArrayWeighted does not have a "rolling" method. Is there any other way I can approach this? Thank you in advance for any help!

**Comment by max-sixty** [comment]
This would be great to have, but unfortunately it's not available in the default methods.

Your approach seems reasonable!

Possibly we could also compute the rolling sum of (values.weights), and divide by the rolling sum of weights...

**Comment by dcherian** [comment, marked as answer]
The answer to not iterating over rolling and coarsen objects is usually `rolling.construct`. This will give you a Xarray object with an "expanded" view of the data and then you can run `weighted` on that. 

This [tutorial](https://tutorial.xarray.dev/intermediate/01-high-level-computation-patterns.html) should be useful.

**Reply by corakingdon** [reply]
Thank you for your answers! I've found a way to achieve this with `construct`, but there is a slightly different result: the first two values are NaNs in my above example, and are not Nans in the following code:
```python
import xarray as xr
import numpy as np

window_size = 3

da = xr.DataArray(np.arange(0.,12.), coords={"time":np.arange("2022-01", "2023-01", dtype='datetime64[M]')})
month_weights = xr.DataArray(da.time.dt.daysinmonth, coords = da.coords) # The numbers of days in each month are the desired weights

da_rolling = da.rolling(time=window_size).construct("window")
month_weights_rolling = month_weights.rolling(time=window_size).construct("window")

da_weighted = da_rolling.weighted(month_weights_rolling.fillna(0))
result2 = da_weighted.mean("window")
```
So the two different resulting arrays are the following:
```
result    # from the previous example using manual iteration
<xarray.DataArray (time: 12)>
array([        nan,         nan,  1.        ,  2.02247191,  3.        ,
        4.        ,  5.        ,  6.01086957,  6.98913043,  8.        ,
        9.        , 10.        ])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01
  
result2    # from the current example using construct
<xarray.DataArray (time: 12)>
array([ 0.        ,  0.47457627,  1.        ,  2.02247191,  3.        ,
        4.        ,  5.        ,  6.01086957,  6.98913043,  8.        ,
        9.        , 10.        ])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01
```

In both cases, the first two values are taking a mean that has data values of Nan but weights of zero (at the start of the rolling windows). But one returns Nans while the second just weights the value as zero. I can see either behavior as making sense, but I am surprised that they are not the same. Do you know why they are different, and should they be?

**Reply by dcherian** [reply]
Ah confusing. When you iterate the first window is `[np.nan]`, the second is `[np.nan, np.nan]`, and the third is `[np.nan, np.nan, 0]`. So for the first two it's a nanmean over all NaN's = NaN.

When you construct, your first row is `[nan, nan,  0.],` so there is a non-NaN value, this is returned by the `nanmean`

This is quite confusing.

**Reply by corakingdon** [reply]
ah okay thank you, so `construct` behavior is not actually the same views as the windows in the rolling iterator. I would like the behavior of returning nans for the first two values. Do you know if there is a way for me to achieve this in my second "construct" method? i.e. is there any option I can specify for taking means of nans to return a nan...?

**Reply by dcherian** [reply]
If you want the mean to return NaN if there is at least one NaN then use `da_rolling.mean("window", skipna=False)`. This will also give you significant memory savings.

If you want to set NaNs when a certain number of NaNs exist in a window use something like `mean('window').where(da_rolling.count("window") > 2)`

**Reply by corakingdon** [reply]
perfect, thank you so much!

**Reply by corakingdon** [reply]
Just to add as a follow-up comment: it appears that the `skipna` kwarg is only used for rolling averages if you first use construct. (is this an issue with the `mean` method of DataArrayRolling objects? it doesn't error, but it appears to ignore it)

```
# add a NAN in the middle of my data:
da[4] = np.NAN

# using `skipna=True` does nothing here
da.rolling(time=3).mean("time", skipna=True)
<xarray.DataArray (time: 12)>
array([nan, nan,  1.,  2., nan, nan, nan,  6.,  7.,  8.,  9., 10.])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01

# same as skipna=False
da.rolling(time=3).mean("time", skipna=False)
<xarray.DataArray (time: 12)>
array([nan, nan,  1.,  2., nan, nan, nan,  6.,  7.,  8.,  9., 10.])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01

# whereas if you call construct first, then the skipna kwarg works
da.rolling(time=3).construct("window").mean("window", skipna=True)
<xarray.DataArray (time: 12)>
array([ 0. ,  0.5,  1. ,  2. ,  2.5,  4. ,  5.5,  6. ,  7. ,  8. ,  9. ,
       10. ])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01

# skipna=False
da.rolling(time=3).construct("window").mean("window", skipna=False)
<xarray.DataArray (time: 12)>
array([nan, nan,  1.,  2., nan, nan, nan,  6.,  7.,  8.,  9., 10.])
Coordinates:
  * time     (time) datetime64[ns] 2022-01-01 2022-02-01 ... 2022-12-01
```

**Reply by dcherian** [reply]
That looks like a bug. Can you open an issue please? 

Note `rolling` has a `min_count` kwarg that also affects this behaviour.

**Reply by corakingdon** [reply]
sure, will start a new issue!

**Reply by corakingdon** [reply]
https://github.com/pydata/xarray/issues/6772
